File: [23.12]Design-Pattern/Java/src/creational/builder/Hamburger.java
package creational.builder;

public class Hamburger {
    private String name;
    private String size;
    private int price;

    public static class HamburgerBuilder{
        private String name;
        private String size;
        private int price;

        public HamburgerBuilder name(String name){
            this.name = name;
            return this;
        }

        public HamburgerBuilder size(String size){
            this.size = size;
            return this;
        }

        public HamburgerBuilder price(int price){
            this.price = price;
            return this;
        }

        public Hamburger build(){
            return new Hamburger(this.name, this.size, this.price);
        }
    }

    private Hamburger(String name, String size, int price) {
        this.name = name;
        this.size = size;
        this.price = price;
    }

    @Override
    public String toString() {
        return "Hamburger{" +
                "name='" + name + '\'' +
                ", size='" + size + '\'' +
                ", price=" + price +
                '}';
    }
}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/creational/builder/Hamburger.kt
package creational.builder

class Hamburger private constructor(
    val name: String,
    val size: String,
    val price: String
) {
    // 중첩 클래스로 정적 이너 클래스와 비슷한 개념이다.
    class Builder(
        private var name: String = "",
        private var size: String = "",
        private var price: String = "",
        ) {
        fun name(name: String): Builder {
            this.name = name
            return this
        }

        fun size(size: String): Builder {
            this.size = size
            return this
        }

        fun price(price: String): Builder {
            this.price = price
            return this
        }

        fun build(): Hamburger {
            return Hamburger(name, size, price)
        }
    }

    override fun toString(): String {
        return "Hamburger(name='$name', size='$size', price='$price')"
    }


}